src/lib/Type.hs

```diff
@@ -140,6 +149,29 @@ instance HasType Atom where
       return ty
     VariantTy row -> checkLabeledRow row $> TyKind
     ACase e alts resultTy -> checkCase e alts resultTy
+    DataConRef def@(DataDef _ paramBs cons) params args -> do
+      checkEq (length paramBs) (length params)
+      forM (zip (toList paramBs) (toList params)) $ \(b, param) ->
+        param |: binderAnn b
+      let argBs' = applyNaryAbs (Abs paramBs argBs) params
+      checkDataConRefBindings argBs' args
+      return $ RawRefTy $ TypeCon def params
+      where DataConDef _ argBs = cons !! 0
```

Comment:
Can we replace cons with `[cons]` instead, just to make sure that the expectation is that there should be only one element?